### ISSUE TITLE
Remove BatchPolopt

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#1391]
This is really a bundle of helper functions masquerading as a superclass. It makes the inheritance tree unnecessarily-deep and obscures implementation details of the algorithms which use it.

We should restrict our use of inheritance to the most obvious/terse relationships (e.g. VPG<--PPO), and not use it to group helper functions shared by a set of classes.

It can be replaced by moving its helper functions (namely, process_samples) to a function package, and moving the main algo loop into each algorithm.

**Comment by naeioi** [comment]
This issue has actually two independent parts - the Numpy BatchPolopt and the TensorFlow BatchPolopt. The numpy one is inherited by both numpy and torch trees. If we were to remove BatchPolopt, we better break it to two issues.

This idea of replacing BatchPolopt with a package of helpers looks tricky to me. There're many accesses to the `self` object. A helper will be much more verbose. We need more discussion on the plan for this issue.

**Comment by ryanjulian** [comment]
`BatchPolopt` is basically
1. Constructor which saves some fields
1. Short `train` loop (and `train_once` in the case of tf)
1. `process_samples`

(1) and (2) can be copied into the subclasses*, increasing clarity a lot. 
(3) can be a helper function.

* - `discount` and `max_path_length` should probably be put into RLAlgorithm

There's no reason one issue can't be addressed by a series of PRs. Feel free to edit it with a checklist, if you want.

**Comment by ryanjulian** [comment]
Let's not overestimate the magnitude here:

```sh
rjulian@nyquist:~/code/garage$ rg BatchPolopt src -l
src/garage/np/algos/__init__.py
src/garage/np/algos/cem.py
src/garage/np/algos/cma_es.py
src/garage/np/algos/batch_polopt.py
src/garage/torch/algos/vpg.py
src/garage/tf/algos/__init__.py
src/garage/tf/algos/reps.py
src/garage/tf/algos/npo.py
src/garage/tf/algos/batch_polopt.py
```

The two versions of this class have 5 usages total.

**Comment by AiRuiChen** [comment]
> `BatchPolopt` is basically
> 
> 1. Constructor which saves some fields
> 2. Short `train` loop (and `train_once` in the case of tf)
> 3. `process_samples`
> 
> (1) and (2) can be copied into the subclasses*, increasing clarity a lot.
> (3) can be a helper function.
> 
> * * `discount` and `max_path_length` should probably be put into RLAlgorithm
> 
> There's no reason one issue can't be addressed by a series of PRs. Feel free to edit it with a checklist, if you want.

1. Why don't we put `env_spec`, `policy`, `baseline`, `discount`, `max_path_length` in RLAlgorithm? These are the common variables of the two versions of BatchPolopt.

2. I agree that since there are many accesses to `self` in `process_samples`, it seems verbose to create a function package and pass all variables to it.  Copying `process_samples` into each algo class sounds like a bad coding practice, as we have to maintain multiple copies of `process_samples`. Any thoughts/recommendations?

**Comment by ryanjulian** [comment]
I think that this refactor is healthy and process_samples can definitely be a function.

It's basically `preprocess_samples_for_policy_gradient`

1.
  * `env_spec`, `discount` are universal to RL algorithms, and should go in `RLAlgorithm`
  * `policy` and `baseline` are specific to on-policy RL algorithms which use policy gradients (inherit from NPO or VPG), so shouldn't be in `RLAlgorithm`
 * `max_path_length` does not have an unambiguous definition (i.e. it is the same for sampling and evaluation in on-policy RL, but those two can differ in off-policy RL). this is really a parameter of sampling/evaluation so probably shouldn't be in `RLAlgorithm`.
2. References to `self` in process_samples:
  * `self.max_path_length` -- pass as parameter (it's used for padding the batch to a fixed size). this won't even be necessary when we transition on-policy algos to TrajectoryBatch
  * `self.baseline` -- these should be computed into values inside the algorithm and passed as an argument (this will happen during the ValueFunction refactor in TF anyway -- see how the torch tree does it). they come from a neural network and we don't want this function responsible for evaluating that neural network.
  * `self.discount` -- should be passed by value. this is a property of RL algorithms
  * `self.episode_reward_mean` this is a statistic that got shoved in here to help someone compare to OpenAI baselines' non-standard performance metrics. if it needs to happen it should be in one of the algorithm classes/superclasses, not this function. we can probably just remove it since we now use AverageDiscountedReturn universally, but @ahtsan added it, so he would know better

Much of this function will be removed when we refactor the on-policy algos to use TrajectoryBatch. There may actually be no point in its existence then, there's no point fussing over how long it is now.

**Comment by ryanjulian** [comment]
@AiRuiChen for context, we want to remove this informal data structure (usually named `paths`) which is used throughout many of the algorithms and interfaces, and replace it will the well-defined `TrajectoryBatch` (see `src/garage/_dtypes.py`).

Most of `process_samples` takes raw data from the sampler and either (1) pads it to a fixed length or (2) aggregates it into a different shape. Neither of those are necessary using the new sampler interface (which produces `TrajectoryBatch`s), so it's not clear this function will even need to exist soon.

**Comment by ryanjulian** [comment]
See for background:
https://github.com/rlworkgroup/garage/pull/1353
https://github.com/rlworkgroup/garage/blob/master/src/garage/_dtypes.py
https://github.com/rlworkgroup/garage/issues/1110
https://github.com/rlworkgroup/garage/issues/1057

In particular, a lot of the functionality in `process_samples` is covered by `TrajectoryBatch.concatenate`:
https://github.com/rlworkgroup/garage/blob/master/src/garage/_dtypes.py#L233
